core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala

```diff
@@ -253,4 +248,32 @@ abstract class KafkaServerTestHarness extends QuorumTestHarness {
     getController().kafkaController.controllerContext.topicNames.toMap
   }
 
+  private def createBrokers(startup: Boolean): Unit = {
+    // Add each broker to `servers` buffer as soon as it is created to ensure that brokers
+    // are shutdown cleanly in tearDown even if a subsequent broker fails to start
+    for (config <- configs) {
+      val broker = createBrokerFromConfig(config)
+      _brokers += broker
+      if (startup) {
+        broker.startup()
```

Comment:
It seems like we should be setting the `alive` boolean to true here, rather than later on, right? Otherwise this might get out of sync if we get an exception later on.

Reply:
This PR is not changing how this works in general, so this comment applies to the pre-PR implementation as well.  I think the assumption is that the list of alive brokers is only valid if **all** brokers get created and are (now optionally) started -- and the test should be failed/torn down if an exception should occur that prevents all brokers from being processed.  I'll add a comment to this effect in `recreateBrokers()` (which is the only place this gets invoked aside from in `setUp()`).

Reply:
It seems simple to just set the boolean right after creating the broker. Why not do that?

I agree this is a pre-existing problem, but that's not a reason to propagate it.

Reply:
It'd debatable if it is a problem.  It is not a problem if indeed the assumption is that the list of alive brokers is only valid if all brokers get created and are (now optionally) started -- if it isn't expected to be valid then putting in the effort to keep it valid doesn't strictly add anything.  That being said, I agree the effort is not onerous to do so, and keeping things valid wherever possible is best in the long-run, so I'll push a commit to fix this.  Thanks for this feedback.